presto-orc/src/main/java/com/facebook/presto/orc/writer/SliceDictionaryColumnWriter.java

```diff
@@ -312,23 +319,66 @@ private void bufferOutputData()
                 presentStream.writeBoolean(dictionaryIndexes.get(position) != 0);
             }
             for (int position = 0; position < rowGroup.getValueCount(); position++) {
-                int dictionaryIndex = dictionaryIndexes.get(position);
+                int originalDictionaryIndex = dictionaryIndexes.get(position);
                 // index zero is reserved for null
-                if (dictionaryIndex != 0) {
-                    int value = dictionaryIndex - 1;
-                    if (value < 0) {
+                if (!dictionaryElements.isNull(originalDictionaryIndex)) {
```

Comment:
Use the same check for `presentStream` (line 319). It is up to you whether you use `!= 0` or `dictionaryElements.isNull`.

I prefer using one for-loop to populate `presentStream` and `dataStream`. In that setup, there won't be a consistent problem. But it's up to you.